# Numpy

In [1]:
import numpy as np

In [2]:
# %timeit  # to time any operation for comparision

In [3]:
a = np.array([1,2,3,4])

In [4]:
type(a)

numpy.ndarray

In [5]:
b = np.array((3,4,5))

In [6]:
print(b)
type(b)

[3 4 5]


numpy.ndarray

In [7]:
x = np.zeros((3,4))
print (x)   #rank od 2 ----2d array
type(x)
x.shape

[[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]]


(3, 4)

In [8]:
y = np.ones((3,4), dtype =np.int16)
print(y)
y.shape

[[1 1 1 1]
 [1 1 1 1]
 [1 1 1 1]]


(3, 4)

In [9]:
z = np.zeros((2,3,4))
print (z)   #rank od 3 ----3d array
type(z)

[[[0. 0. 0. 0.]
  [0. 0. 0. 0.]
  [0. 0. 0. 0.]]

 [[0. 0. 0. 0.]
  [0. 0. 0. 0.]
  [0. 0. 0. 0.]]]


numpy.ndarray

In [10]:
y = np.ones((3,4), dtype =np.float64)
print(y)

[[1. 1. 1. 1.]
 [1. 1. 1. 1.]
 [1. 1. 1. 1.]]


In [11]:
z = np.full((3,4), 101)
print(z)

[[101 101 101 101]
 [101 101 101 101]
 [101 101 101 101]]


In [12]:
pi = np.full((3,4), np.pi)
print(pi)

[[3.14159265 3.14159265 3.14159265 3.14159265]
 [3.14159265 3.14159265 3.14159265 3.14159265]
 [3.14159265 3.14159265 3.14159265 3.14159265]]


In [13]:
#np.arange ------Starts with till last number
np.arange(10,30)

array([10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26,
       27, 28, 29])

In [14]:
np.arange(10,30, 4)

array([10, 14, 18, 22, 26])

In [15]:
np.arange(0,1.1, 0.2)

array([0. , 0.2, 0.4, 0.6, 0.8, 1. ])

In [16]:
x = [
    [
        [1],
        [1],
        [1],
        [1]],
    [
        [1],
        [1],
        [1],
        [1]]
]

In [17]:
xx= np.array(x)
xx.shape

(2, 4, 1)

In [18]:
#linear space
np.linspace(0, 100, 11)

array([  0.,  10.,  20.,  30.,  40.,  50.,  60.,  70.,  80.,  90., 100.])

In [19]:
np.linspace(1, 10, 25)

array([ 1.   ,  1.375,  1.75 ,  2.125,  2.5  ,  2.875,  3.25 ,  3.625,
        4.   ,  4.375,  4.75 ,  5.125,  5.5  ,  5.875,  6.25 ,  6.625,
        7.   ,  7.375,  7.75 ,  8.125,  8.5  ,  8.875,  9.25 ,  9.625,
       10.   ])

In [20]:
x= np.random.rand(2,3)
print(x, type(x))

[[0.64785026 0.8361103  0.82929816]
 [0.49855817 0.03706656 0.61639333]] <class 'numpy.ndarray'>


In [21]:
x = np.empty((2,3))

In [22]:
x

array([[0.64785026, 0.8361103 , 0.82929816],
       [0.49855817, 0.03706656, 0.61639333]])

In [23]:
c = np.arange(1,5)
c

array([1, 2, 3, 4])

In [24]:
c.dtype

dtype('int64')

In [25]:
c.itemsize

8

In [26]:
a = np.arange(6)

In [27]:
a

array([0, 1, 2, 3, 4, 5])

In [28]:
b = a.reshape(2,3)
print(b)

[[0 1 2]
 [3 4 5]]


In [29]:
b = a.reshape(2,2)
print(b)

ValueError: cannot reshape array of size 6 into shape (2,2)

In [ ]:
a = np.arange(1,15,2)
print (a)

In [ ]:
a[4]

In [ ]:
print(a[2:5])

In [30]:
print(a[2::2])
print(a[2::3])

[2 4]
[2 5]


In [31]:
print(a[::-1])

[5 4 3 2 1 0]


In [32]:
a[2:5] = -1
print(a)

[ 0  1 -1 -1 -1  5]


In [33]:
a[:] = 0
print(a)

[0 0 0 0 0 0]


In [34]:
# Carefull update
a = np.array ([1,3,6,9,12,15,18,21,22])
print(a)

a_slice = a[4:]
print(a_slice)

a_slice[1:5]=100
print(a_slice)

print(a)

[ 1  3  6  9 12 15 18 21 22]
[12 15 18 21 22]
[ 12 100 100 100 100]
[  1   3   6   9  12 100 100 100 100]


In [35]:
a = np.array ([1,3,6,9,12,15,18,21,22])
print(a)

a_slice = a[4:].copy()   #copy
print(a_slice)

a_slice[1:5]=100
print(a_slice)   #update

print(a)         #unchanged

[ 1  3  6  9 12 15 18 21 22]
[12 15 18 21 22]
[ 12 100 100 100 100]
[ 1  3  6  9 12 15 18 21 22]


In [36]:
x = np.random.rand(4,5)
print(x)

[[0.30204737 0.506353   0.58220849 0.08043013 0.13881144]
 [0.30187111 0.32009235 0.20501048 0.22506782 0.83010496]
 [0.72875671 0.897249   0.98259622 0.91329984 0.25246407]
 [0.94913853 0.68350839 0.10427169 0.70590022 0.99547475]]


In [37]:
x[1]     # row

array([0.30187111, 0.32009235, 0.20501048, 0.22506782, 0.83010496])

In [38]:
x[1, 0]    #row/columns

0.30187110847230836

In [39]:
x [0:1, 0:4,]  # [ Row:Row , Column: Column]

array([[0.30204737, 0.506353  , 0.58220849, 0.08043013]])

In [40]:
a = np.arange(12).reshape(3,4)
print (a)

row_on = np.array([True, False, True])

print(row_on)

a[row_on]

[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]
[ True False  True]


array([[ 0,  1,  2,  3],
       [ 8,  9, 10, 11]])

In [41]:
a = np.array([20, 30, 40 , 50])
print(a)
b= np.arange(4)
print(b)
c = a+b
print(c)

[20 30 40 50]
[0 1 2 3]
[20 31 42 53]


In [42]:
import numpy as np
A = np.array ([[1,1],
              [0,1]])
B = np.array ([[2,0],
              [3,4]])

print(A*B)     # element wise product


print('------------------------------------')

print (([[1*2,1*0],
              [0*3,1*4]]))

[[2 0]
 [0 4]]
------------------------------------
[[2, 0], [0, 4]]


In [43]:

print(np.dot(A,B))

print('------------------------------------')

print (([[1*2+1*3,1*0+1*4],[0*2+1*3, 0*0+1*4]]))

[[5 4]
 [3 4]]
------------------------------------
[[5, 4], [3, 4]]


In [44]:
# Vector is defined by magnitude and direction
# vector = list of scalars

# Using vectors is machine learning

# X =  f1  f2  f3  f4
#   = [10, 12, 23, 40] -------Features as vectror

# Glove Vectors
# Queen - women + man = King
# relationship between Man & Women has same with King & Queen


In [45]:
# numpy.array([1, 1, 1, 1])

##### Broadcasting in Numpy Array

##### 2 rules of broadcasting

###### 1st rule
First rule of Broadcasting

[[[1, 3 ]]] + [5] [[[6, 8]]]

Shape (1, 1, 2) (1, ) (1, 1, 2)

If the arrays do not have the same rank, then a 1 will be
prepended to the smaller ranking arrays until their ranks match.



In [46]:
a = np.arange(6).reshape(2,3)
print("A = ",a)
b = np.array([1])
print("B = ", b)
c = a + b
print(c)

A =  [[0 1 2]
 [3 4 5]]
B =  [1]
[[1 2 3]
 [4 5 6]]


In [47]:
c = a-b
print(c)

[[-1  0  1]
 [ 2  3  4]]


In [48]:
a = np.array([[1]])
print("A = ",a, "Shape = ",a.shape)
b = np.array([1])
print("B = ", b, "Shape = ",b.shape)
c = a+b
print("C = " , c,"Shape = ", c.shape)

A =  [[1]] Shape =  (1, 1)
B =  [1] Shape =  (1,)
C =  [[2]] Shape =  (1, 1)


##### 2nd rule
<img src="https://s3.amazonaws.com/cloudxlab/static/images/course/numpy_pandas_for_ml/Broacasting_2_rule.png" alt="drawing" width="500"/>

In [49]:
k = np.arange(6).reshape(2, 3)
k

k + [100, 200, 300]


array([[100, 201, 302],
       [103, 204, 305]])

In [50]:
# Python array
a - [1,2,3,4,5]
a.mean()

1.0

In [51]:
#numpy array
a = np.array([[-2.5, 3.1, 7], [10, 11, 12]])  # sum of all elements/count
print("mean =", a.mean())

mean = 6.766666666666667


Similar to mean there are other ndarray methods which can be
used for various computations.

* min - returns the minimum element in the ndarray
* max - returns the maximum element in the ndarray
* sum - returns the sum of the elements in the ndarray
* prod - returns the product of the elements in the ndarray
* std - returns the standard deviation of the elements in the ndarray.
* var - returns the variance of the elements in the ndarray.

In [52]:
a = np.array([[-2.5, 3.1, 7], [10, 11, 12]])

for func in (a.min, a.max, a.sum, a.prod, a.std,a.var):
    print(func.__name__, "=", func())

min = -2.5
max = 12.0
sum = 40.6
prod = -71610.0
std = 5.084835843520964
var = 25.855555555555554


In [53]:
c=np.arange(24).reshape(2,3,4)
c

array([[[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]],

       [[12, 13, 14, 15],
        [16, 17, 18, 19],
        [20, 21, 22, 23]]])

In [54]:
c.sum(axis=0) # sum across matrices

array([[12, 14, 16, 18],
       [20, 22, 24, 26],
       [28, 30, 32, 34]])

In [55]:
# Important about numpy sum
a= np.array ([[1,2,3],
             [4,5,6]])
b= np.array ([[.1,.2,.3],
             [.4,.5,.6]])

In [56]:
a+b

array([[1.1, 2.2, 3.3],
       [4.4, 5.5, 6.6]])

In [57]:
a= np.array ([[1,2,3],
             [4,5,6]])
a.sum()

21

In [58]:
a.sum(axis = 0)

array([5, 7, 9])

In [59]:
a.sum(axis = 1)

array([ 6, 15])

In [60]:
# Transpose
>>> m1 = np.arange(6).reshape(2,3)
>>> m1
>>> m1.T

array([[0, 3],
       [1, 4],
       [2, 5]])

#### Solving a system of linear scalar equations
2x + 6y = 6
5x + 3y = -9

In [61]:
from numpy import linalg

In [62]:
coeffs = np.array([[2, 6], [5, 3]]) #2x + 6y = 6 5x + 3y = -9
depvars = np.array([6, -9])
solution = linalg.solve(coeffs, depvars)
solution

array([-3.,  2.])

#### Solving a system of linear scalar equations

In [63]:
coeffs.dot(solution), depvars

(array([ 6., -9.]), array([ 6, -9]))

<div class="panel-body">
<div class="aha-panel-body">
                    
<h3>Creating Numpy array by passing a Python list</h3>

<p>Please follow the below steps to create a NumPy array by passing a Python list:</p>

<ul>
<li><p>Import Numpy as np</p>

<pre><code>import numpy as np
</code></pre></li>
<li><p>Create a Python list as below:</p>

<pre><code>sample_list = [1,2,3]
</code></pre></li>
<li><p>Pass the above list to <code>array()</code> function of NumPy</p>

<pre><code>list_array = np.array(&lt;&lt;your code comes here&gt;&gt;)
</code></pre></li>
</ul>

<h3>Creating Numpy array by passing a Python tuple</h3>

<p>Please follow the below steps to create a NumPy array by passing a Python tuple:</p>

<ul>
<li><p>Import Numpy as np</p>

<pre><code>import numpy as np
</code></pre></li>
<li><p>Create a Python tuple as below:</p>

<pre><code>tup = (1,2,3)
</code></pre></li>
<li><p>Pass the above tuple <code>tup</code>  to <code>array()</code> function of NumPy</p>

<pre><code>my_tup_array = np.array(&lt;&lt;your code comes here&gt;&gt;)
</code></pre></li>
</ul>

<p>You can use the <code>print()</code> function to view the above-created arrays.</p>

 </div>
</div>

In [64]:
import numpy as np

In [65]:
my_list = [50, 12, 67, 23]

In [66]:
my_list_array = np.array(my_list)

In [67]:
print(my_list_array)

[50 12 67 23]


In [68]:
my_tup = (88, 17, 45, 76)

In [69]:
my_tup_arr = np.array(my_tup)

In [70]:
print(my_tup_arr)

[88 17 45 76]


In [71]:
import numpy as np

In [72]:
sample_list = [1,2,3]

In [73]:
list_array = np.array(sample_list)

In [74]:
tup = (1,2,3)

In [75]:
my_tup_array = np.array(tup)

<div class="panel-body">
<div class="aha-panel-body">
                    
<h3>Time comparison of a NumPy operation and a corresponding Python code</h3>

<p>Let us create the below function (<code>multiply_loops</code>) which takes two arrays as input and computes their <strong>multiplication using normal Python way</strong>.</p>

<pre><code>def multiply_loops(A, B):
    c=np.zeros((A.shape[0], B.shape[1]))
    for i in range(A.shape[0]):
        for k in range(B.shape[1]):
            c[i,k] = 0
            for j in range(B.shape[0]):
                n = A[i,j] * B[j,k]
                c[i,k] += n
    return c
</code></pre>

<p>Now, let us create the below function (<code>multiply_vector</code>) which takes two arrays as input and computes their <strong>multiplication using NumPy's vector multiplication way</strong>.</p>

<pre><code>def multiply_vector(A, B):
     return A @ B
</code></pre>

<p>Let us create two randomly generated 100x100 matrices - <code>X</code> and <code>Y</code> - to test the above functions</p>

<pre><code>X = np.random.random((100, 100))
Y = np.random.random((100, 100))
</code></pre>

<p>Now execute the below command (<code>timeit</code>) in Jupyter, which will output you the time taken by each of these functions</p>

<pre><code>%timeit multiply_loops(X, Y)
%timeit multiply_vector(X, Y)
</code></pre>

                    
</div>
 </div>

In [76]:
def multiply_loops(A, B):
    c=np.zeros((A.shape[0], B.shape[1]))
    for i in range(A.shape[0]):
        for k in range(B.shape[1]):
            c[i,k] = 0
            for j in range(B.shape[0]):
                n = A[i,j] * B[j,k]
                c[i,k] += n
    return c

In [77]:
def multiply_vector(A, B):
     return A @ B
    

In [78]:
X = np.random.random((100, 100))
Y = np.random.random((100, 100))

In [79]:
print(X)

[[0.24152173 0.95906566 0.99791582 ... 0.52826371 0.99504092 0.30887238]
 [0.72533448 0.32554667 0.79260992 ... 0.45182862 0.15062709 0.07299403]
 [0.53222206 0.70859501 0.13654581 ... 0.15748053 0.75570108 0.3378911 ]
 ...
 [0.00349584 0.86283961 0.79165897 ... 0.56212922 0.49618213 0.602017  ]
 [0.69430327 0.5050416  0.71743691 ... 0.20018207 0.50985696 0.86974961]
 [0.95810923 0.89476795 0.73905254 ... 0.13242955 0.13654371 0.07416635]]


In [80]:
print(Y)

[[0.38280934 0.9295094  0.51072163 ... 0.07129882 0.53279332 0.89371929]
 [0.79891556 0.41229234 0.11700565 ... 0.83794956 0.10337687 0.82369528]
 [0.43661118 0.02328076 0.09557493 ... 0.28532287 0.67983302 0.77648558]
 ...
 [0.47228628 0.63472244 0.23533187 ... 0.49457907 0.98333012 0.99795726]
 [0.79169989 0.8228851  0.1749444  ... 0.46431455 0.31185185 0.41004487]
 [0.31686387 0.71163836 0.68288719 ... 0.99572315 0.0349002  0.27446441]]


In [81]:
%timeit multiply_loops(X, Y)


865 ms ± 14.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [82]:
%timeit multiply_vector(X, Y)

65.9 µs ± 6.85 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)


<div class="aha-panel-body">
                    
<p><strong>Note:</strong> Please import numpy as np</p>

<p><strong>(1) Creating NumPy array with all its elements values as zero (using <code>zeros()</code> function of NumPy)</strong></p>

<ul>
<li><p>Create a tuple indicating dimensions of the desired array</p>

<pre><code>tup_dim = (3, 4)
</code></pre></li>
<li><p>Now, create your NumPy array by passing the above tuple (<code>tup_dim</code>) to <code>zeros()</code> function below</p>

<pre><code>my_zeros_array = np.zeros(&lt;&lt;your code comes here&gt;&gt;, dtype=np.int16)
</code></pre></li>
</ul>

<p><strong>(2) Creating NumPy array with all its elements values as one (using <code>ones()</code> function of NumPy)</strong></p>

<ul>
<li><p>Create a tuple indicating dimensions of the desired array</p>

<pre><code>tup_dim = (3, 4)
</code></pre></li>
<li><p>Now, create your NumPy array by passing the above tuple (tup_dim) to <code>ones()</code> function below</p>

<pre><code>my_ones_array = np.ones( &lt;&lt;your code comes here&gt;&gt;, dtype=np.int16)
</code></pre></li>
</ul>

<p><strong>(3) Creating NumPy array with all its elements values as a specific given value (using <code>full()</code> function of NumPy)</strong></p>

<ul>
<li><p>Create a tuple indicating dimensions of the desired array</p>

<pre><code>tup_dim = (3, 4)
</code></pre></li>
<li><p>Now, create your NumPy array by passing the above tuple (<code>tup_dim</code>) to <code>full()</code> function below, along with the desired value that you want the array to be filled with (e.g. say value 7)</p>

<pre><code>my_seven_array = np.full( &lt;&lt;your code comes here&gt;&gt;, 7, dtype=np.int16)
</code></pre></li>
</ul>

<p>The above NumPy array <code>my_seven_array</code> will be filled with value 7.</p>

<p><strong>(4) Creating Identity matrix or array</strong></p>

<p>Identity matrix (array) is a square matrix with all its elements as zero (0) except for the diagonal elements whose value is one (1). That is, all the main diagonal values are 1s (one).</p>

<p>You can create an Identity matrix by using numpy's <code>identity()</code> function by passing the desired dimension of the matrix (array) and the data type required.</p>

<p>e.g. np.identity(2, dtype=float) will create a 2x2 matrix with all its values as 0.0 except for the diagonal values which will be 1.0</p>

<ul>
<li><p>Please create an identity matrix <code>my_identity_array</code> of size 4x4 with float values.</p>

<pre><code>my_identity_array = np.identity(&lt;&lt;your code comes here&gt;&gt;, dtype=np.float64)
</code></pre></li>
</ul>

<p>You can use the print() function to view the above-created arrays.</p>

<p>NOTE: <strong>Please ensure that <code>dtype</code> of <code>my_zeros_array, my_ones_array, my_seven_array</code> is <code>np.int16</code> and <code>dtype</code> of <code>my_identity_array</code> is <code>np.float64</code>.</strong></p>

                    
                </div>

In [83]:
tup_dim = (3, 4)

my_zeros_array = np.zeros(tup_dim, dtype = np.int16)

In [84]:
type(my_zeros_array)

numpy.ndarray

In [85]:
tup_dim = (3, 4)
my_ones_array = np.ones(tup_dim, dtype=np.int16)

In [86]:
my_ones_array

array([[1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1]], dtype=int16)

In [87]:
tup_dim = (3, 4)
my_seven_array = np.full( tup_dim, 7, dtype=np.int16)

In [88]:
my_seven_array

array([[7, 7, 7, 7],
       [7, 7, 7, 7],
       [7, 7, 7, 7]], dtype=int16)

In [89]:
my_identity_array = np.identity(4, dtype=np.float64)

In [90]:
my_identity_array

array([[1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.]])

<div class="panel panel-default">
            <div class="panel-heading">
                <strong>
                    INSTRUCTIONS
                </strong>
            </div>
            <div class="panel-body">
                <div class="aha-panel-body">
<p><strong>Note:</strong> Please import numpy as np</p>

<p><strong>(1) NumPy array with all its elements values as random values (using <code>random.rand()</code> function of NumPy)</strong></p>

<ul>
<li><p>Create your NumPy array by passing the above the desired dimensions of the array (say (3,4))  to <code>random.rand()</code> function below</p>

<pre><code>my_rand_array = np.random.&lt;&lt;your code comes here&gt;&gt;(3,4)
</code></pre></li>
</ul>

<p>The above NumPy array <code>my_rand_array</code> will have random float values between 0.0 and 1.0</p>

<p><strong>(2) NumPy array with all its elements values as uninitialized values (initialized with some random values which you are not aware of - using the <code>empty()</code> function of NumPy)</strong></p>

<ul>
<li><p>Create a tuple indicating dimensions of the desired array</p>

<pre><code>tup_dim = (3, 4)
</code></pre></li>
<li><p>Now, create your NumPy array by passing the above tuple (<code>tup_dim</code>) to <code>empty()</code> function below</p>

<pre><code>my_uninitialized_array = np.empty( &lt;&lt;your code comes here&gt;&gt;)
</code></pre></li>
</ul>

<p>The above NumPy array <code>my_uninitialized_array</code> will be filled with random (also unknown) float values between 0.0 and 1.0</p>

                    
</div>
</div>
</div>

In [91]:
my_rand_array = np.random.rand(3,4)

In [92]:
my_rand_array

array([[0.28831978, 0.28140446, 0.88434241, 0.36913981],
       [0.7553149 , 0.77118054, 0.359725  , 0.94526554],
       [0.94124214, 0.41025977, 0.05647742, 0.77661137]])

In [93]:
tup_dim = (3, 4)

In [94]:
my_uninitialized_array = np.empty(tup_dim)

In [95]:
my_uninitialized_array

array([[0.28831978, 0.28140446, 0.88434241, 0.36913981],
       [0.7553149 , 0.77118054, 0.359725  , 0.94526554],
       [0.94124214, 0.41025977, 0.05647742, 0.77661137]])

<div class="panel panel-default">
            <div class="panel-heading">
                <strong>
                    INSTRUCTIONS
                </strong>
            </div>
            <div class="panel-body">
                <div class="aha-panel-body">
                    
<p><strong>Note:</strong> Please import numpy as np</p>

<p><strong>(1)  Creating a NumPy array by passing a range of values (using <code>arange()</code> function of NumPy)</strong></p>

<p>Suppose you want to create a NumPy array whose values should lie between 10 and 30 and the values  should have a gap of the value of 5</p>

<ul>
<li><p>To create this NumPy array, pass the start value (10), end value (30) and the gap value 5, to the  <code>arange()</code> function below</p>

<pre><code>my_range_array = np.arange( &lt;&lt;your code comes here&gt;&gt;)
</code></pre></li>
</ul>

<p>The above NumPy array my_range_array will be filled with values between 10 and 30 (excluding 30) having a gap of the value of 5 between the consecutive values.</p>

<p><strong>(2) Creating a NumPy array by passing an equally spaced range of values (using <code>linspace()</code> function of NumPy)</strong></p>

<p>Suppose you want to create a NumPy array whose values should lie between 0 and 5/3 and the consecutive values should have an equal amount of gap between them (equally spaced values).</p>

<ul>
<li><p>To create this NumPy array, pass the start value (0), end value (5/3), and the total number of values you want (say 6),  to the <code>linspace()</code> function below</p>

<pre><code>my_spaced_array = np.linspace( &lt;&lt;your code comes here&gt;&gt;)
</code></pre></li>
</ul>

<p>The above NumPy array <code>my_spaced_array</code> will have total of 6 values with values ranging between 0 and 5/3 (including 5/3) and having an equal amount of gap between the consecutive values.</p>

                    
</div>
</div>
</div>

In [96]:
my_range_array = np.arange(10,30,5)
my_range_array

array([10, 15, 20, 25])

In [97]:
my_spaced_array = np.linspace(0,5/3,6)
my_spaced_array

array([0.        , 0.33333333, 0.66666667, 1.        , 1.33333333,
       1.66666667])

<div class="aha-panel-body">
                    <h1 class="h2">
                        NumPy - Arrays - Attributes of a NumPy Array
                    </h1>
                    
<p>NumPy array (ndarray class) is the most used construct of NumPy in Machine Learning and Deep Learning. Let us look into some important attributes of this NumPy array.</p>

<p>Let us create a Numpy array first, say, <code>array_A</code>.</p>

<p>Pass the above list to <code>array()</code> function of NumPy</p>

<pre><code>array_A = np.array([ [3,4,6], [0,8,1] ])
</code></pre>

<p>Now, let us understand some important attributes of ndarray object using the above-created array <code>array_A</code>.</p>

<p><strong>(1) ndarray.ndim</strong></p>

<p><code>ndim</code> represents the number of dimensions (axes) of the ndarray.</p>

<p>e.g. for this 2-dimensional array [ [3,4,6], [0,8,1]], value of <code>ndim</code> will be 2. 
This ndarray has two dimensions (axes) - rows (axis=0) and columns (axis=1)</p>

<p><strong>(2) ndarray.shape</strong></p>

<p><code>shape</code> is a tuple of integers representing the size of the ndarray in each dimension.</p>

<p>e.g. for this 2-dimensional array [ [3,4,6], [0,8,1]], value of <code>shape</code> will be (2,3)
because this ndarray has two dimensions - rows and columns - and the number of rows is 2 and the number of columns is 3</p>

<p><strong>(3) ndarray.size</strong></p>

<p><code>size</code> is the total number of elements in the ndarray. It is equal to the product of elements of the shape. 
e.g. for this 2-dimensional array [ [3,4,6], [0,8,1]], <code>shape</code> is (2,3), <code>size</code> will be product (multiplication) of 2 and 3 i.e. (2*3) = 6. Hence, the size is 6.</p>

<p><strong>(4) ndarray.dtype</strong></p>

<p><code>dtype</code> tells the data type of the elements of a NumPy array. In NumPy array, all the elements have the same data type.</p>

<p>e.g. for this NumPy array [ [3,4,6], [0,8,1]], <code>dtype</code> will be <code>int64</code></p>

<p><strong>(5) ndarray.itemsize</strong></p>

<p><code>itemsize</code> returns the size (in bytes) of each element of a NumPy array. </p>

<p>e.g. for this NumPy array [ [3,4,6], [0,8,1]], <code>itemsize</code> will be 8, because this array consists of integers and size of integer (in bytes) is 8 bytes.</p>

                    
</div>

<div class="row">
    <div class="col-md-12">
        <div class="panel panel-default">
            <div class="panel-heading">
                <strong>
                    INSTRUCTIONS
                </strong>
            </div>
            <div class="panel-body">
                <div class="aha-panel-body">
                    
<p>Please follow the below steps:</p>

<p>(1) Please <code>import numpy as np</code></p>

<p>(2) Create a NumPy array with name <code>my_array</code> with the below elements</p>

<pre><code>[ [1, 4, 5, 6], [7, 8, 9, 10], [11, 12, 14, 16] ]
</code></pre>

<p>Now, based on the above-created array (<code>my_array</code>), please answer the questions in the next few slides.</p>

 </div>
            </div>
        </div>
    </div>
</div>

In [98]:
my_array = np.array([ [1, 4, 5, 6], [7, 8, 9, 10], [11, 12, 14, 16] ])
my_array

array([[ 1,  4,  5,  6],
       [ 7,  8,  9, 10],
       [11, 12, 14, 16]])

In [99]:
# What is the dimension of NumPy array my_array ?
my_array.ndim

2

In [100]:
# What is the shape of NumPy array my_array ?
my_array.shape


(3, 4)

In [101]:
# What is the data type of elements of NumPy array my_array?
my_array.dtype


dtype('int64')

In [102]:
# What is the size of NumPy array my_array?
my_array.size


12

In [103]:
# What is the itemsize of NumPy array my_array?
my_array.itemsize

8

<div class="aha-panel-body">
					<h1 class="h2">
						Numpy - Arrays - Loading a text file data using NumPy's loadtxt() function  - Step 1
					</h1>
					
<p>Although Pandas provides better ways and constructs to load a dataset from various sources like files, databases, etc. we should also know the NumPy constructs for the same.</p>

<p>There are two ways (constructs) in NumPy to load data from a text file:</p>

<p>(1) using <code>loadtxt()</code> function</p>

<p>(2) using <code>genfromtxt()</code> function</p>

<p><code>loadtxt()</code> function provides less flexibility, whereas <code>genfromtxt()</code> function provides more flexibility.</p>

<p>For example, <code>genfromtxt()</code> function also handles the missing values kind of scenarios in the loaded dataset, whereas <code>loadtxt()</code> function doesn't.</p>

<h3>Example of loadtxt()</h3>

<pre><code>import numpy as np
name_arr, address_arr, zipcode_arr = np.loadtxt('my_file.txt', skiprows=2, unpack=True)
</code></pre>

<p>The above piece of code will load the data from <code>my_file.txt</code> text file. </p>

<p><code>skiprows=2</code> means, skip the first two rows of the <code>my_file.txt</code> file while loading the data.</p>

<p><code>unpack=True</code> means, unpack the columns of the dataset being loaded and return the data of each column separately in separate arrays ( name column data in <code>name_arr</code> array,  address column data in <code>address_arr</code> array,  zipcode column data in <code>zipcode_arr</code> array).</p>

<p><code>unpack=False</code> means, return only a single array as output from the <code>loadtxt()</code> function.</p>

					
</div>

<div class="aha-panel-body">
                    
<p>Please follow the below steps:</p>

<h3>(1) Import the required libraries</h3>

<pre><code>import numpy as np
import os
</code></pre>

<h3>(2) Load using pandas</h3>

<p>Now we will use pandas to load data from a large csv file (California housing dataset) and create a small csv file (of housing data) by extracting only few rows of data from this large  housing.csv file.</p>

<p>We are creating a smaller csv file of data, just for our convenience, to make it easy for us to load it using loadtxt() function.</p>

<p>Don't worry if you don't know pandas yet, just copy and use the below pandas code as it is.</p>

<pre><code>import pandas as pd
# defining housing.csv file path
HOUSING_PATH =  '/cxldata/datasets/project/housing'
# reading the large housing.csv file using pandas
housing_raw = pd.read_csv(os.path.join(HOUSING_PATH, "housing.csv"))
# extracting only a few rows (5 rows) of data from the pandas dataframe 'my_df'
my_df = housing_raw.iloc[ : 5]
# creating a new small csv file - 'housing_short.csv' - containing the above extracted 5 rows of data
my_df.to_csv('housing_short.csv', index=False)
</code></pre>

<h3>(3) Load using Numpy</h3>

<p>Now, let us load the csv file - <code>housing_short.csv</code> - using NumPy's <code>loadtxt()</code> function</p>

<p>please define a variable called FILE and assign to it the string value <code>housing_short.csv</code>.</p>

<pre><code>FILE = '&lt;&lt;your code comes here&gt;&gt;'
</code></pre>

<h3>(4) Create Function</h3>

<p>Please define a function called <code>load_housing_data()</code>, as shown below, which takes filename (FILE) as input and loads this file using NumPy's <code>loadtxt()</code> function. Just copy the below code as it is.</p>

<pre><code>def load_housing_data(file = FILE ):
    return np.loadtxt(file, dtype={'names': ('longitude','latitude','housing_median_age','total_rooms','total_bedrooms','population','households','median_income','median_house_value','ocean_proximity'),'formats': ('f8', 'f8', 'f8', 'f8', 'f8', 'f8', 'f8', 'f8', 'f8', '|S15')}, delimiter=',', skiprows=1, unpack=True)
</code></pre>

<h3>loadtxt() function parameters</h3>

<p><strong>first parameter</strong> - file. It is the name of the file from which the data is to be loaded.</p>

<p><strong>second parameter</strong> - data type <code>dtype</code> of columns of the loaded csv file <code>housing_short.csv</code>. It is a Python dictionary with <em>key as <code>names</code></em> of the columns, and  <em><code>values</code> as the data types</em> of these respective columns e.g. f8, |S15, etc.</p>

<p>'f8' means  64-bit floating-point number</p>

<p>'|S15' -means a string of length of 15 characters</p>

<p><strong>third parameter</strong> - delimiter. It is the character by which values in a row of our csv file are separated. For example, in our case values of a row of our csv file - <code>housing_short.csv</code> - are separated by ',' (comma)</p>

<p><strong>fourth parameter</strong> - skiprows. You can specify here, how many initial rows of the csv file you want to skip loading. E.g. you may want to skip the first row of this csv file, as it may contain header information in the first row, which you may not want to load.</p>

<p><strong>fifth parameter</strong> - unpack. When unpack is True, the returned array is transposed, so that arguments may be unpacked using <code>x, y, z = loadtxt(...)</code>. When used with a structured data-type, arrays are returned for each field. The default value for unpack is False. But here we are returning the individual arrays so we have kept it here asTrue.</p>

<h3>(5) Call the Function</h3>

<p>Please call the above defined <code>load_housing_data()</code> function, which returns various column values as NumPy arrays</p>

<pre><code>longitude_arr,latitude_arr,housing_median_age_arr,total_rooms_arr,total_bedrooms_arr,population_arr,households_arr,median_income_arr,median_house_value_arr,ocean_proximity_arr = load_housing_data()
</code></pre>

<h3>(6) Print</h3>

<p>You can just check and confirm the values of one of the NumPy arrays (say <code>median_house_value_arr</code>) that you got above by printing the same using <code>print()</code> function</p>

<pre><code>print(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p><code>median_house_value_arr</code> contains values of median_house_value column of the csv file - <code>housing_short.csv</code></p>

                    
</div>

In [104]:
import numpy as np
import os

In [105]:
import pandas as pd
# defining housing.csv file path
HOUSING_PATH =  '/cxldata/datasets/project/housing'

# reading the large housing.csv file using pandas
housing_raw = pd.read_csv(os.path.join(HOUSING_PATH, "housing.csv"))

# extracting only a few rows (5 rows) of data from the pandas dataframe 'my_df'
my_df = housing_raw.iloc[ : 5]

# creating a new small csv file - 'housing_short.csv' - containing the above extracted 5 rows of data
my_df.to_csv('housing_short.csv', index=False)

In [106]:
FILE = 'housing_short.csv'

In [107]:
def load_housing_data(file = FILE ):
    return np.loadtxt(file, dtype={'names': 
                                   ('longitude','latitude','housing_median_age','total_rooms','total_bedrooms','population','households','median_income','median_house_value','ocean_proximity'),
                                   'formats': ('f8', 'f8', 'f8', 'f8', 'f8', 'f8', 'f8', 'f8', 'f8', '|S15')},
                                    delimiter=',', skiprows=1, unpack=True)

In [108]:
longitude_arr,latitude_arr,housing_median_age_arr,total_rooms_arr,total_bedrooms_arr,population_arr,households_arr,median_income_arr,median_house_value_arr,ocean_proximity_arr = load_housing_data()

In [109]:
print(median_house_value_arr)

[452600. 358500. 352100. 341300. 342200.]


<div class="aha-panel-body">
                    <h1 class="h2">
                        Numpy - Arrays - Loading a text file data using NumPy's genfromtxt() function
                    </h1>
                    
<p>As we discussed earlier, there are two ways (constructs) in NumPy to load data from a text file:</p>

<p>(1) using <code>loadtxt()</code> function</p>

<p>(2) using <code>genfromtxt()</code> function</p>

<p><br></p>

<p>Below is an example of using <code>genfromtxt()</code> function</p>

<p><br></p>

<h3>Example of <code>genfromtxt()</code></h3>

<p><br></p>

<p><code>genfromtxt()</code> function is very helpful when you are expecting some missing values in the dataset to be loaded. Below is a sample code </p>

<pre><code>import numpy as np
my_arr = np.genfromtxt('my_file.txt', skip_header=2, filling_values=9999999)
</code></pre>

<p>Here, if all your data in the dataset is of type integer then, by default, the string values are treated as missing values, and <code>genfromtxt()</code> function will replace these missing values (string values) with a <code>nan</code> value.</p>

<p>If you want the missing values to be replaced with some other value other than <code>nan</code>, then, you can specify this particular value in the <code>filling_values</code> parameter. For example, in the above code, we are saying that if any missing values found, please replace it with value 9999999.</p>

<p><code>genfromtxt()</code> function also trims any white spaces around the values being loaded. </p>

<p>You can also specify if you want to load any maximum number of rows, in this case, only specified number of max. rows will be loaded.</p>

                    
</div>

<div class="aha-panel-body">
                    
<h3>Please follow the below steps:</h3>

<p>(1) Please import the required libraries</p>

<pre><code>import numpy as np
import os
</code></pre>

<p>(2) Please create a variable HOUSING_PATH and assign to it the path of <code>housing.csv</code> file 
(<code>'/cxldata/datasets/project/housing'</code>) as a string</p>

<pre><code>HOUSING_PATH = &lt;&lt;your code comes here&gt;&gt;
</code></pre>

<p>(3) Please define a complete path for your csv file <code>housing.csv</code> by using <code>os.path.join()</code> function, by passing to it the HOUSING_PATH and the csv file <code>housing.csv</code>, and save this complete path in a variable FILE.</p>

<pre><code>FILE = os.path.join(HOUSING_PATH, &lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p>(4) Please define a function <code>load_housing_dataset()</code> and add to it the complete path of the csv file (FILE) just defined above. This function will load <code>housing.csv</code> file using <code>genfromtxt()</code> function.</p>

<pre><code>def &lt;&lt;your code comes here&gt;&gt;(file =FILE ):
    return np.genfromtxt(file, dtype={'names': ('longitude','latitude','housing_median_age','total_rooms','total_bedrooms','population','households','median_income','median_house_value','ocean_proximity'),'formats': ('f8', 'f8', 'f8', 'f8', 'f8', 'f8', 'f8', 'f8', 'f8', '|S15')}, delimiter=',', skip_header=1, filling_values = 99999999, unpack=False)
</code></pre>

<p><strong><code>genfromtxt()</code> function parameters:</strong></p>

<p><strong>first parameter</strong> - name of the file from which the data is to be loaded.</p>

<p><strong>second parameter</strong> - data type (dtype) of columns of the loaded csv file <code>housing.csv</code>. It is a Python dictionary with key as 'names' of the columns, and  'values' as the data types of these respective columns e.g. f8, |S15, etc.</p>

<p>'f8' means  64-bit floating-point number
'|S15' -means a string of length of 15 characters</p>

<p><strong>third parameter</strong> - delimiter. Character by which values in a row of our csv file are separated. For example, in our case values of a row of our csv file  <code>housing.csv</code>  are separated by ',' (comma)</p>

<p><strong>fourth parameter</strong> - skiprows. You can specify here, how many initial rows of the csv file you want to skip loading. E.g. you may want to skip the first row of this csv file, as it may contain header information in the first row, which you may not want to load.</p>

<p><strong>fifth parameter</strong> - unpack. Same meaning as explained in <code>loadtxt()</code> function chapter.</p>

<p>(5) Call the <code>load_housing_dataset()</code> function, as defined above, and store the output in a variable called result_arr</p>

<pre><code>result_arr = &lt;&lt;your code comes here&gt;&gt;()
</code></pre>

<p>(6) Print the length (number of records) of <code>result_arr</code></p>

<pre><code>print(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p>(7) Print array <code>result_arr</code> to see its values.</p>

<pre><code>print(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

                    
 </div>

In [110]:
import numpy as np
import os

In [111]:
HOUSING_PATH = '/cxldata/datasets/project/housing'

In [112]:
FILE = os.path.join(HOUSING_PATH, 'housing.csv')

In [113]:
def load_housing_dataset(file =FILE ):
    return np.genfromtxt(file, dtype={'names': ('longitude','latitude','housing_median_age','total_rooms','total_bedrooms','population','households','median_income','median_house_value','ocean_proximity'),'formats': ('f8', 'f8', 'f8', 'f8', 'f8', 'f8', 'f8', 'f8', 'f8', '|S15')}, delimiter=',', skip_header=1, filling_values = 99999999, unpack=False)

In [114]:
result_arr = load_housing_dataset()

In [115]:
print(result_arr.size)

20640


In [116]:
print(result_arr)

[(-122.23, 37.88, 41.,  880.,  129.,  322.,  126., 8.3252, 452600., b'NEAR BAY')
 (-122.22, 37.86, 21., 7099., 1106., 2401., 1138., 8.3014, 358500., b'NEAR BAY')
 (-122.24, 37.85, 52., 1467.,  190.,  496.,  177., 7.2574, 352100., b'NEAR BAY')
 ...
 (-121.22, 39.43, 17., 2254.,  485., 1007.,  433., 1.7   ,  92300., b'INLAND')
 (-121.32, 39.43, 18., 1860.,  409.,  741.,  349., 1.8672,  84700., b'INLAND')
 (-121.24, 39.37, 16., 2785.,  616., 1387.,  530., 2.3886,  89400., b'INLAND')]


### NumPy - Arrays - Resizing an Array
Resize()
<div class="aha-panel-body">
                    
<p>Please follow the below-mentioned steps:</p>

<p>(1) Please <code>import numpy as np</code></p>

<p>(2) Please create a (2X4) numpy array (i.e. with 2 rows and 4 columns)  called <code>my_arr</code> with values - 1,2,3,4,5,6,7 and 8,  using <code>array()</code> function of numpy library </p>

<pre><code>&lt;&lt;your code comes here&gt;&gt; = np.array(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p>(3) Please apply <code>resize()</code> function on this <code>my_arr</code> array to convert this (2X4) array into a (3x4) array.</p>

<pre><code>my_arr.resize(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p>(4) Please print this array (<code>my_arr</code>) to see its values</p>

                    
</div>

In [117]:
my_arr = np.array([[1,2,3,4],[5,6,7,8]])

In [118]:
my_arr.resize(3,4)

In [119]:
print(my_arr)

[[1 2 3 4]
 [5 6 7 8]
 [0 0 0 0]]


<h3>Difference between resize() and reshape() :</h3>
<p><strong>reshape()</strong> will create an array with the same number of elements as the original array, i.e. of the same ‘size’ as that of the original array. If you want to convert the original array to a bigger array, <code>reshape()</code> can’t add more elements(than available in the original array) to give you a bigger array. Hence, <code>reshape()</code> is used to create a new array of the same size (as the original array) but of different desired dimensions.</p>

<p><strong>resize()</strong> can create an array of larger size than the original array. To convert the original array into a bigger array, <code>resize()</code> will add more elements (than available in the original array) by copying the existing elements (repeated as many times as required) to fill the desired larger size. <code>resize()</code> is used to create a new array of different sizes and dimensions.</p>

<div class="aha-panel-body">
                    
<p>Please follow the below-mentioned steps:</p>

<p><br>
<strong>First case:</strong> When, for the reshaped array,  the number of desired rows and columns are known.
<br>
<br>
(1) Please <code>import numpy as np</code></p>

<p>(2) Please create a 1-D numpy array called <code>my_first_arr</code> with values - 1,2,3,4,5,6,7 and 8,  using <code>array()</code> function of numpy library </p>

<pre><code>&lt;&lt;your code comes here&gt;&gt; = np.array(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p>(3) Please apply <code>reshape()</code> function on this <code>my_first_arr</code> array to convert this 1-D array into a (2x4) array and store this newly created (2x4) array in a variable called <code>my_new_arr</code></p>

<pre><code>&lt;&lt;your code comes here&gt;&gt; = my_first_arr.reshape(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p>(4) Please print this new array (<code>my_new_arr</code>) to see its values</p>

<p>(5) Please print the existing array (<code>my_first_arr</code>) to see its values</p>

<p><br>
<strong>Second case:</strong> When, for the reshaped array,  the number of desired columns is known, but the desired number of rows is not known.
<br>
<br>
(1) Please <code>import numpy as np</code></p>

<p>(2) Please create a 1-D numpy array called <code>my_second_arr</code> with values - 0,1,2,3,4,5,6,7 and 8,  using <code>array()</code> function of numpy library </p>

<pre><code>&lt;&lt;your code comes here&gt;&gt; = np.array(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p>(3) Please apply <code>reshape()</code> function on this <code>my_second_arr</code> array to convert this 1-D array into an array with 3 columns but unknown number of rows, and store this newly created array in a variable called <code>my_updated_arr</code></p>

<pre><code>&lt;&lt;your code comes here&gt;&gt; = my_second_arr.reshape(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p>(4) Please print this new array (<code>my_updated_arr</code>) to see its values</p>

<p>(5) Please print the existing array (<code>my_second_arr</code>) to see its values</p>

                    
</div>

In [120]:
my_first_arr = np.array([1,2,3,4,5,6,7,8])

In [121]:
my_new_arr = my_first_arr.reshape(2,4)

In [122]:
print(my_new_arr)

[[1 2 3 4]
 [5 6 7 8]]


In [123]:
print(my_first_arr)

[1 2 3 4 5 6 7 8]


In [124]:
# Second case
my_second_arr = np.array([0,1,2,3,4,5,6,7,8])

In [125]:
my_updated_arr = my_second_arr.reshape(-1, 3)   #unknown rows

In [126]:
print(my_updated_arr)

[[0 1 2]
 [3 4 5]
 [6 7 8]]


In [127]:
print(my_second_arr)

[0 1 2 3 4 5 6 7 8]


### Numpy - Arrays - Indexing and Array Slicing
<div class="aha-panel-body">
                    
<p>Please follow the below instructions to try the below problems using array index slicing technique:</p>

<pre><code>Please import numpy as np
</code></pre>

<p><br></p>

<h3>Reversing of an array elements</h3>

<p><br>
<br>
(1) Please define a NumPy array with name <code>my_orig_arr</code> having following elements (1, 15, 3, 9, 26, 7, 89, 12) </p>

<pre><code>my_orig_arr = np.array(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p>(2) Please reverse the elements of the above array by using the below syntax and store the result in a variable called <code>my_rev_arr</code></p>

<pre><code>&lt;&lt; your code comes here&gt;&gt; = my_orig_arr[ : : -1]
</code></pre>

<p>(3) Now print this array <code>my_rev_arr</code> that you got above to see its values</p>

<pre><code>print(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p><br></p>

<h3>Replacing selected elements of a NumPy array by a specific value</h3>

<p><br>
<br></p>

<ul>
<li><p>Create a new array <code>my_orig_arr2</code> having following elements (2, 9, 17, 13, 1, 4, 20, 57).</p>

<pre><code>my_orig_arr2 = np.array(&lt;&lt;your code comes here&gt;&gt;)
</code></pre></li>
<li><p>Say, you want to replace a few elements (say from index 2 to 5) of the above NumPy array <code>my_orig_arr2</code> by a value <code>-6</code>, you can do so by following the below steps:</p>

<pre><code>my_orig_arr2[2 : 5] = -6
</code></pre></li>
<li><p>Please print this NumPy array <code>my_orig_arr2</code> to see its values (please note that only values at indexes 2, 3, 4 will be replaced, not the 5th index value)</p>

<pre><code>print(&lt;&lt;your code comes here&gt;&gt;)
</code></pre></li>
</ul>

<p><strong>Please remember that this is only possible in a NumPy array, if you try to do this with a normal Python array, you will get an error.</strong></p>

<p>e.g.</p>

<p>(1) Please create a normal Python array</p>

<pre><code>arr = [1, 3, 5, 10, 4, 15]
</code></pre>

<p><code>arr</code> is a normal Python array here. </p>

<p>(2) Please try to replace some of its values with a value say <code>-3</code>, you will get an error</p>

<pre><code>arr[1 : 4]  = -3
</code></pre>

<p><br></p>

<h3>Extracting a slice(portion) of a NumPy array</h3>

<p><br></p>

<p>To extract a slice (portion) of a NumPy array, please follow the below steps:</p>

<p>(1) Define a NumPy array with name <code>apple</code> with values (1, 8, 23, 3, 18, 91, 7, 15)</p>

<pre><code>apple = np.array(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p>(2) Extract a portion of this NumPy array  <code>apple</code> from index 1 to 4 (by using the below syntax) and store it in a variable <code>apple_slice</code></p>

<pre><code>&lt;&lt;your code comes here&gt;&gt; = apple[1 : 4]
</code></pre>

<p>(3) Print array <code>apple_slice</code> using the <code>print()</code> function to see its values</p>

<pre><code>print(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p>(4) Print array <code>apple</code> using the <code>print()</code> function to see its values are intact</p>

<pre><code>print(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p><strong>If you change (replace) value in this slice (<code>apple_slice</code> array), the corresponding value will also get modified in its parent array <code>apple</code>.</strong> </p>

<p>Please follow the below steps to check this for yourself.</p>

<p>(1) Replace 2nd element of array <code>apple_slice</code> by a value say <code>99999</code></p>

<pre><code>apple_slice[1] = 99999
</code></pre>

<p>please remember 2nd element will have an index of '<code>1</code>'</p>

<p>(3) Print array <code>apple_slice</code> using the <code>print()</code> function to see its values, 2nd element should be 99999 now</p>

<pre><code>print(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p>(4) Print array <code>apple</code> using the <code>print()</code> function to see its values, the corresponding original value in the parent array <code>apple</code> should also have been changed to <code>99999</code></p>

<pre><code>print(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p><strong>Original (parent) array from which the slice was extracted also gets modified if the slice value is modified.</strong></p>

<p><strong>If you don't want original (parent) array values to get modified when its slice values are changing, you can create the slice by using a <code>copy()</code> function in the slicing code syntax</strong></p>

<ul>
<li>Suppose you don't want your parent (original) array values to get modified when its slice values are changing. You can achieve this by using the <code>copy()</code> function in the previous code syntax to create a slice.</li>
</ul>

<p>Please follow the below steps:</p>

<p>(1) Please create a new slice from the <code>apple</code> array by using <code>copy()</code> function in the syntax, as shown below:</p>

<pre><code>apple_slice_new = apple[2 : 5].copy()
</code></pre>

<p>(2) Replace 2nd element of array <code>apple_slice_new</code> by a value say <code>222222</code></p>

<pre><code>apple_slice_new[1] = 222222
</code></pre>

<p>please remember 2nd element will have an index of '<code>1</code>'</p>

<p>(3) Print slice array <code>apple_slice_new</code> using the <code>print()</code> function to see its values, 2nd element should be <code>222222</code> now</p>

<pre><code>print(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p>(4) Print parent array <code>apple</code> using the <code>print()</code> function to see its values, the parent array <code>apple</code> should be intact (no values modified).</p>

<pre><code>print(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p><strong>Original (parent) array from which the slice was extracted using the <code>copy()</code> function doesn't get modified if the slice value is modified.</strong></p>

                    
</div>

In [128]:
my_orig_arr = np.array([1, 15, 3, 9, 26, 7, 89, 12])

In [129]:
# Reversing of an array elements
my_rev_arr = my_orig_arr[ : : -1]
print(my_rev_arr)

[12 89  7 26  9  3 15  1]


In [130]:
#Replacing selected elements of a NumPy array by a specific value
my_orig_arr2 = np.array([2, 9, 17, 13, 1, 4, 20, 57])
my_orig_arr2[2 : 5] = -6
print(my_orig_arr2)

[ 2  9 -6 -6 -6  4 20 57]


In [131]:
# Please remember that this is only possible in a NumPy array, 
# if you try to do this with a normal Python array, you will get an error.
# arr = [1, 3, 5, 10, 4, 15]
# arr[1 : 4]  = -3
import numpy as np

In [132]:
# Extracting a slice(portion) of a NumPy array
apple = np.array([1, 8, 23, 3, 18, 91, 7, 15])
apple_slice = apple[1 : 4]
print(apple_slice)
print(apple)

[ 8 23  3]
[ 1  8 23  3 18 91  7 15]


**** If you change (replace) value in this slice (apple_slice array), the corresponding value will also get modified in its parent array apple. ****

In [133]:
apple_slice[1] = 99999
print(apple_slice)

[    8 99999     3]


In [134]:
print(apple)

[    1     8 99999     3    18    91     7    15]


Original (parent) array from which the slice was extracted also gets modified if the slice value is modified.

If you don't want original (parent) array values to get modified when its slice values are changing, you can create the slice by using a copy() function in the slicing code syntax

In [135]:
apple_slice_new = apple[2 : 5].copy()
apple_slice_new[1] = 222222
print(apple_slice_new)

[ 99999 222222     18]


In [136]:
print(apple)

[    1     8 99999     3    18    91     7    15]


Original (parent) array from which the slice was extracted using the copy() function doesn't get modified if the slice value is modified.


#### Numpy - Arrays - Example - Extracting a portion of an image using array slicing

<div class="aha-panel-body">
                    
<p>Please follow the below steps:</p>

<p>(1) Please import the required libraries</p>

<pre><code>import numpy as np
from sklearn.datasets import load_sample_image
</code></pre>

<p>(2) Let us use the <code>load_sample_images</code> function to load <code>china.jpg</code> image and store the loaded image in a variable called <code>china</code></p>

<pre><code>china = load_sample_image("china.jpg")
</code></pre>

<p>(3) Please check the shape of <code>china</code> array</p>

<pre><code>china.&lt;&lt;your code comes here&gt;&gt;
</code></pre>

<p>the shape of <code>china</code> array is <code>(427, 640, 3)</code></p>

<p>Let us see the structure of this <code>china</code> array and then we will try to understand each value in this shape.</p>

<p><code>china</code> is a 3-dimensional numpy array, and it looks something like below</p>

<pre><code>[
      [ [174, 201, 231], [250, 251, 255],...........]    - row 0
      [ [172, 199, 229], [251, 252, 255],...........]    - row 1
      .
      .
]
</code></pre>

<p>Hence, in <code>china</code> array, we see that 
<br></p>

<ul>
<li>there are some rows - the second level brackets ('<code>[</code>') represents rows. Like this, we have <strong>427</strong> rows.</li>
<li>each row has some pixels (each sub-array like <code>[174, 201, 231]</code> in the row represents a pixel). Like this we have <strong>640</strong> pixels in each row.</li>
<li>the <strong>3</strong> values inside each pixel array ( [<code>174, 201, 231</code>] ) represent colour codes for the colour of this pixel. e.g. 174, 201 and 231 combinedly may be representing a dark brown colour (colour of this pixel of the china image).</li>
</ul>

<p><br>
If you see the values above (marked in bold font), you will know what each value of the shape (427, 640, 3) represents in the china array.</p>

<p>(4) Just for simplifying the things, let us extract a small portion of this <code>china</code> numpy array and store it in a variable called <code>image</code>.
Here, we are extracting (slicing) a portion (<code>image</code>) of <code>china</code> array from row number 150 to 220 and column number 130 to 250.</p>

<pre><code>image = china[150:220, 130:250]
</code></pre>

<p>Now, let us check the shape of this <code>image</code> array by using the below code</p>

<pre><code>image.&lt;&lt;your code comes here&gt;&gt;
</code></pre>

<p>(5) Please store each value of shape 'tuple' separately in variables - <code>height</code>, <code>width</code> and <code>channels</code></p>

<pre><code>height, width, channels = image.shape
</code></pre>

<p>Here, <code>channels</code> are nothing but the number of colour codes i.e. 3.</p>

<p>(6) Let us convert the coloured <code>image</code> into a grayscale image (<code>image_grayscale</code>), by replacing the values which are along the <code>axis=2</code>, by their mean value.</p>

<p>Do you remember the concept of axis explained earlier? 
<br></p>

<ul>
<li><code>axis=0</code> represents the first dimension of an array, i.e. it represents the row</li>
<li><code>axis=1</code> represents the second dimension of an array, i.e. it represents the column</li>
<li><code>axis=2</code> represents the third dimension of an array, in this case, the third dimension is <code>channels</code>.
<br></li>
</ul>

<p><code>channels</code> (<code>axis=2</code>) has 3 values (e.g. [68, 39, 44] ), and, mean of these 3 values  is 50.33
Hence, value [68, 39, 44] will be replaced with value 50.33 after the step - <code>image.mean(axis=2)</code></p>

<pre><code>   image_grayscale = image.mean(axis=2).astype(np.float32)
</code></pre>

<p>Now, let us check the shape of this <code>image_grayscale</code> array by using the below code</p>

<pre><code>image_grayscale.&lt;&lt;your code comes here&gt;&gt;
</code></pre>

<p>(7) Now, let us reshape this <code>image_grayscale</code> array into a 4-dimensional array (from existing 2-dimensions) and store the output in a variable called <code>images</code>.</p>

<pre><code>&lt;&lt;your code comes here&gt;&gt; = image_grayscale.reshape(1, height, width, 1)
</code></pre>

<p>(8) Please check dimensions of <code>images</code> array</p>

<pre><code>images.&lt;&lt;your code comes here&gt;&gt;
</code></pre>

                    
</div>

In [137]:
#(1) Please import the required libraries
from sklearn.datasets import load_sample_image

In [138]:
#(2) Let us use the load_sample_images function to load china.jpg image and store the loaded image in a variable called china
china = load_sample_image("china.jpg")
china.shape

(427, 640, 3)

In [139]:
print(china)

[[[174 201 231]
  [174 201 231]
  [174 201 231]
  ...
  [250 251 255]
  [250 251 255]
  [250 251 255]]

 [[172 199 229]
  [173 200 230]
  [173 200 230]
  ...
  [251 252 255]
  [251 252 255]
  [251 252 255]]

 [[174 201 231]
  [174 201 231]
  [174 201 231]
  ...
  [252 253 255]
  [252 253 255]
  [252 253 255]]

 ...

 [[ 88  80   7]
  [147 138  69]
  [122 116  38]
  ...
  [ 39  42  33]
  [  8  14   2]
  [  6  12   0]]

 [[122 112  41]
  [129 120  53]
  [118 112  36]
  ...
  [  9  12   3]
  [  9  15   3]
  [ 16  24   9]]

 [[116 103  35]
  [104  93  31]
  [108 102  28]
  ...
  [ 43  49  39]
  [ 13  21   6]
  [ 15  24   7]]]


In [140]:
# (4) Just for simplifying the things, let us extract a small portion of this china numpy array and store it in a variable called image. 
# Here, we are extracting (slicing) a portion (image) of china array from row number 150 to 220 and column number 130 to 250.
image = china[ 150:220, 130:250]

In [141]:
# Now, let us check the shape of this image array by using the below code
image.shape

(70, 120, 3)

In [142]:
# (5) Please store each value of shape 'tuple' separately in variables - height, width and channels
height, width, channels = image.shape

In [143]:
# (6) Let us convert the coloured image into a grayscale image (image_grayscale), by replacing the values which are along the axis=2, by their mean value.
image_grayscale = image.mean(axis=2).astype(np.float32)

In [144]:
# Now, let us check the shape of this image_grayscale array by using the below code
image_grayscale.shape

(70, 120)

In [145]:
# (7) Now, let us reshape this image_grayscale array into a 4-dimensional array (from existing 2-dimensions) and store the output in a variable called images.
images = image_grayscale.reshape(1, height, width, 1)

In [146]:
# (8) Please check dimensions of images array
images.ndim

4

#### Numpy - Arrays - Multi-Dimensional Arrays and Boolean Indexing


<div class="aha-panel-body">
                    
<p>Please follow the below steps:</p>

<p>(1) Import numpy as np</p>

<p>(2) Create a 4x5 (4 rows, 5 columns) NumPy array called <code>my_multi_arr</code> </p>

<pre><code>my_multi_arr = np.arange(20).reshape(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p>(3) Extract values from row index numbers 2 to 4 and from column index numbers 2 to 5, and store it in a variable called <code>my_multi_arr_portion</code></p>

<pre><code>my_multi_arr_portion = my_multi_arr[&lt;&lt;your code comes here&gt;&gt;]
</code></pre>

<p>(4) Print the <code>my_multi_arr_portion</code> array using <code>print()</code> function to see its values</p>

<pre><code>print(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

                    
</div>

In [147]:
my_multi_arr = np.arange(20).reshape(4,5)
print(my_multi_arr)

[[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]
 [15 16 17 18 19]]


In [148]:
my_multi_arr_portion = my_multi_arr[2:4, 2:5]
print(my_multi_arr_portion)

[[12 13 14]
 [17 18 19]]


#### Numpy - Flattening a NumPy Array (converting to 1-D array) using ravel() function
If you have a multi-dimensional Numpy array, say, a 3-D, 2-D, etc. array, you can convert it to a 1-dimensional (1-D) array using Numpy function ravel().

You may need to do this, as some of the libraries need the data to be fed in 1-dimensional form.
<p>                                                                         </p>

<div class="aha-panel-body">
                    
<p>Please follow the below steps:</p>

<p>(1) Import numpy as np</p>

<p>(2) Please create a 2-dimensional Numpy array (of size 4x5) called <code>my_2d_arr</code> as below</p>

<pre><code>my_2d_arr = np.arange(20).reshape(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p>(3) Call <code>ravel()</code> function on the above Numpy array (<code>my_2d_arr</code>) to create a 1-dimensional array, and store this 1-D array in a variable called <code>my_1d_arr</code></p>

<pre><code>&lt;&lt;your code comes here&gt;&gt; = my_2d_arr.ravel()
</code></pre>

<p>(4) Print this <code>my_1d_arr</code> array by using <code>print()</code> function to see its value</p>

<pre><code>print(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

                    
</div>

In [149]:
my_2d_arr = np.arange(20).reshape(4,5)
print(my_2d_arr)

[[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]
 [15 16 17 18 19]]


In [150]:
my_1d_arr = my_2d_arr.ravel()
print(my_1d_arr)

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19]


#### Numpy - Mathematical Operations on NumPy Arrays - Addition and Subtraction
<p>                                        

</p>
<div class="aha-panel-body">
                    
<p>Please follow the following steps:</p>

<p>(1) Import the required libraries</p>

<pre><code>import numpy as np
</code></pre>

<p><strong>Addition (element-wise)</strong></p>

<p>(2) Please create a NumPy array <code>a_arr</code> with values (60, 70, 80, 90)</p>

<pre><code>a_arr = np.array(&lt;&lt; your code comes here&gt;&gt;)
b_arr = np.arange(4)
</code></pre>

<p>Now, we have two NumPy arrays - <code>a_arr</code> and <code>b_arr</code></p>

<p>(3) Add these two NumPy arrays (<code>a_arr</code> and  <code>b_arr</code>), and store the result in a variable <code>c_arr</code></p>

<pre><code>&lt;&lt;your code comes here&gt;&gt; = a_arr  + b_arr
</code></pre>

<p>(4) Print array <code>c_arr</code> to see its values</p>

<pre><code>print(c_arr)
</code></pre>

<p><strong>Subtraction (element-wise)</strong></p>

<p>(1) Please create a NumPy array <code>d_arr</code> with values (60, 70, 80, 90)</p>

<pre><code>d_arr = np.array(&lt;&lt; your code comes here&gt;&gt;)
e_arr = np.arange(4)
</code></pre>

<p>Now, we have two NumPy arrays - <code>d_arr</code> and <code>e_arr</code></p>

<p>(2) Subtract NumPy array <code>e_arr</code> from array <code>d_arr</code>, and store the result in a variable called <code>f_arr</code></p>

<pre><code>&lt;&lt;your code comes here&gt;&gt; = d_arr  - e_arr
</code></pre>

<p>(3) Print array <code>f_arr</code> to see its values</p>

<pre><code>print(f_arr)
</code></pre>

                    
</div>

In [151]:
a_arr = np.array([60, 70, 80, 90])
b_arr = np.arange(4)

print(a_arr, '\n', b_arr)

[60 70 80 90] 
 [0 1 2 3]


In [152]:
c_arr = a_arr  + b_arr
print(c_arr)

[60 71 82 93]


In [153]:
d_arr = np.array([60, 70, 80, 90])
e_arr = np.arange(4)
print(d_arr, '\n', e_arr)

[60 70 80 90] 
 [0 1 2 3]


In [173]:
f_arr = d_arr  - e_arr
print(f_arr)

[60 69 78 87]


#### Numpy - Mathematical Operations on NumPy Arrays - Multiplication and Dot Product

<div class="aha-panel-body">
                    
<p>Please follow the below steps:</p>

<p>(1) Please import required libraries</p>

<pre><code> import numpy as np
</code></pre>

<p><strong>Multiplication (element-wise product)</strong></p>

<p>(2) Create two NumPy arrays - <code>A_arr</code> and <code>B_arr</code> - as shown below</p>

<pre><code>A_arr = np.array([ [ 5,9], [4, 7] ])
B_arr = np.array( [ [2, 8], [1, 6] ] )
</code></pre>

<p>(3) Multiply the above two NumPy arrays (<code>A_arr</code> and <code>B_arr</code>) and store the result in a variable <code>M_arr</code></p>

<pre><code>&lt;&lt;your code comes here&gt;&gt; = A_arr  &lt;&lt; your code comes here&gt;&gt; B_arr
</code></pre>

<p>(4) Print the array <code>M_arr</code> to see its values</p>

<pre><code>print(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p><strong>Dot Product (Matrix Product)</strong></p>

<p>(1) Create two NumPy arrays - <code>C_arr</code> and <code>D_arr</code> - as shown below</p>

<pre><code>C_arr = np.array([ [ 5,9], [4, 7] ])
D_arr = np.array( [ [2, 8], [1, 6] ] )
</code></pre>

<p>(2) Perform dot product of the above two NumPy arrays (<code>C_arr</code> and <code>D_arr</code>) and store the result in a variable <code>P_arr</code></p>

<pre><code>&lt;&lt;your code comes here&gt;&gt; = np.dot(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p>(3) Print the array <code>P_arr</code> to see its values</p>

<pre><code>print(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

                    
</div>

In [190]:
 import numpy as np

In [193]:
A_arr = np.array([ [ 5,9], [4, 7] ])
B_arr = np.array( [ [2, 8], [1, 6] ] )
print(A_arr, '\n', B_arr)

[[5 9]
 [4 7]] 
 [[2 8]
 [1 6]]


In [194]:
M_arr = A_arr  * B_arr
print(M_arr)

[[10 72]
 [ 4 42]]


In [198]:
#Dot Product (Matrix Product)
C_arr = np.array([ [ 5,9], [4, 7] ])
D_arr = np.array( [ [2, 8], [1, 6] ] )
print(C_arr, '\n', D_arr)

[[5 9]
 [4 7]] 
 [[2 8]
 [1 6]]


In [199]:
P_arr = np.dot(C_arr, D_arr)
print(P_arr)

[[19 94]
 [15 74]]


#### Numpy - Mathematical Operations on NumPy Arrays - Division, Modulus


<div class="aha-panel-body">
                    
<p>
    
Please follow the below steps:</p>

<p>(1) Please import numpy as np</p>

<p><strong>Division</strong></p>

<p>(2) Please create two NumPy arrays - <code>R_div</code> and <code>S_div</code> - as shown below</p>

<pre><code>R_div = np.array([ [ 25,65], [40, 14] ])
S_div = np.array([ [2, 10], [4, 5] ] )
</code></pre>

<p>(3) Divide (using normal division) above created NumPy array <code>R_div</code> by <code>S_div</code> and store the result in a variable <code>T_div</code></p>

<pre><code>&lt;&lt;your code comes here&gt;&gt; = R_div  &lt;&lt; your code comes here&gt;&gt; S_div
</code></pre>

<p>(4) Print the array <code>T_div</code> to see its values</p>

<pre><code>print(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p><strong>Integer Division</strong></p>

<p>(1) Please import numpy as np</p>

<p>(2) Please create two NumPy arrays - <code>R_int_div</code> and <code>S_int_div</code> - as shown below</p>

<pre><code>R_int_div = np.array([ [ 25,65], [40, 70] ])
S_int_div = np.array([ [2, 10], [8, 3] ] )
</code></pre>

<p>(3) Divide (using integer division) above created NumPy array R_int_div by S_int_div and store the result in a variable <code>T_int_div</code></p>

<pre><code>&lt;&lt;your code comes here&gt;&gt; = R_int_div  &lt;&lt; your code comes here&gt;&gt; S_int_div
</code></pre>

<p>(4) Print the array <code>T_int_div</code> to see its values</p>

<pre><code>print(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p><strong>Modulus</strong></p>

<p>(1) Please import numpy as np</p>

<p>(2) Please create two NumPy arrays - <code>R_mod</code> and <code>S_mod</code> - as shown below</p>

<pre><code>R_mod = np.array([ [ 20,65], [40, 70] ])
S_mod = np.array([ [2, 10], [8, 3] ] )
</code></pre>

<p>(3) Apply modulus operation between above created NumPy arrays <code>R_mod</code> and <code>S_mod</code> and store the result in a variable <code>T_mod</code></p>

<pre><code>&lt;&lt;your code comes here&gt;&gt; = R_mod  &lt;&lt; your code comes here&gt;&gt; S_mod
</code></pre>

<p>(4) Print the array <code>T_mod</code> to see its values</p>

<pre><code>print(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

                    
</div>

In [216]:
R_div = np.array([ [ 25,65], [40, 14] ])
S_div = np.array([ [2, 10], [4, 5] ] )

print(R_div, "\n", S_div)

[[25 65]
 [40 14]] 
 [[ 2 10]
 [ 4  5]]


In [217]:
T_div = R_div  / S_div
print(T_div)

[[12.5  6.5]
 [10.   2.8]]


In [220]:
#Integer Division
R_int_div = np.array([ [ 25,65], [40, 70] ])
S_int_div = np.array([ [2, 10], [8, 3] ] )
print(R_int_div, "\n", S_int_div)

[[25 65]
 [40 70]] 
 [[ 2 10]
 [ 8  3]]


In [222]:
T_int_div = R_div  // S_div
print(T_int_div)

[[12  6]
 [10  2]]


In [223]:
# Modulus
R_mod = np.array([ [ 20,65], [40, 70] ])
S_mod = np.array([ [2, 10], [8, 3] ] )
print(R_mod, "\n", S_mod)

[[20 65]
 [40 70]] 
 [[ 2 10]
 [ 8  3]]


In [224]:
T_mod = R_mod  % S_mod
print(T_mod)

[[0 5]
 [0 1]]


#### Numpy - Mathematical Operations on NumPy Arrays - Exponents

<div class="aha-panel-body">
                    
<p>
    
   
   Please follow the below steps:</p>

<p>(1) Please import the required libraries</p>

<pre><code>import numpy as np
</code></pre>

<p>(2) Please create two NumPy arrays - <code>R_exp</code> and <code>S_exp</code> - as shown below</p>

<pre><code>R_exp = np.array([ [ 4, 2], [3, 4 ] ] )
S_exp = np.array( [ [2, 10], [4, 5] ] )
</code></pre>

<p>(3) Apply the exponent operator <code>**</code> between the above created NumPy arrays <code>R_exp</code> and <code>S_exp</code> and store the result in a variable <code>T_exp</code></p>

<pre><code>&lt;&lt;your code comes here&gt;&gt; = R_exp  &lt;&lt; your code comes here&gt;&gt; S_exp
</code></pre>

<p>(4) Print the array <code>T_exp</code> to see its values</p>

<pre><code>print(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

</div>

In [243]:
R_exp = np.array([ [ 4, 2], [3, 4 ] ] )
S_exp = np.array( [ [2, 10], [4, 5] ] )
print(R_exp, "\n", S_exp)

[[4 2]
 [3 4]] 
 [[ 2 10]
 [ 4  5]]


In [244]:
T_exp = R_exp  ** S_exp
print(T_exp)

[[  16 1024]
 [  81 1024]]


#### Numpy - Mathematical Operations on NumPy Arrays - Conditional Operators
<div class="aha-panel-body">
                    
<p>
    Please follow the below steps:</p>

<p>(1) Please import the required libraries</p>

<pre><code>import numpy as np
</code></pre>

<p>(2) Please create a NumPy array called <code>U</code>, as shown below</p>

<pre><code>U = np.array([ [ 22, 45], [90, 4 ] ] )
</code></pre>

<p>(3)  Get all the elements of <code>U</code> which are less than <code>45</code>, and print the result using the <code>print()</code> function</p>

<pre><code>print( &lt;&lt; your code comes here&gt;&gt; )
</code></pre>

                    
 </div>

In [253]:
U = np.array([ [ 22, 45], [90, 4 ] ] )

In [254]:
print(U < 45)

[[ True False]
 [False  True]]


#### Numpy - Mathematical and Statistical functions on NumPy Arrays

<div class="aha-panel-body">
                    
<p>
    Please follow the below steps:</p>

<p>(1) Please import the required libraries</p>

<pre><code>import numpy as np
</code></pre>

<p>(2) Please create a NumPy array <code>X_stat</code> with values [ [ 1, 2, 3], [ 4, -5, 6] ]</p>

<pre><code>&lt;&lt; your code comes here &gt;&gt; = np.array( &lt;&lt; your code comes here &gt;&gt; )
</code></pre>

<p><strong>mean()</strong></p>

<p>(3) Please calculate and print the mean of elements of this array <code>X_stat</code> </p>

<pre><code>print("mean = ", &lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p><strong>var()</strong></p>

<p>(4) Please calculate and print the variance of elements of this array <code>X_stat</code> </p>

<pre><code>print("Variance = ", &lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p><strong>std()</strong></p>

<p>(5) Please calculate and print the standard deviation of elements of this array <code>X_stat</code> </p>

<pre><code>print("Standard Deviation = ", &lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p><strong>min()</strong></p>

<p>(6) Please calculate and print the min of elements of this array <code>X_stat</code> </p>

<pre><code>print("min = ", &lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p><strong>max()</strong></p>

<p>(7) Please calculate and print the max of elements of this array <code>X_stat</code> </p>

<pre><code>print("max = ", &lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p><strong>sum()</strong></p>

<p>(8) Please calculate and print the sum of elements of this array <code>X_stat</code> </p>

<pre><code>print("sum = ", &lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p><strong>prod()</strong></p>

<p>(9) Please calculate and print the product of elements of this array <code>X_stat</code> </p>

<pre><code>print("product = ", &lt;&lt;your code comes here&gt;&gt;)
</code></pre>

                    
</div>

In [261]:
X_stat = np.array([ [ 1, 2, 3], [ 4, -5, 6] ])
print(X_stat)

[[ 1  2  3]
 [ 4 -5  6]]


In [269]:
# mean()
print("mean = ", X_stat.mean())

mean =  1.8333333333333333


In [270]:
# var()
print("Variance = ", X_stat.var())

Variance =  11.805555555555557


In [271]:
# std()
print("Standard Deviation = ", X_stat.std())

Standard Deviation =  3.435921354681384


In [272]:
# min()
print("min = ", X_stat.min())

min =  -5


In [273]:
# max()
print("max = ", X_stat.max())

max =  6


In [274]:
# sum()
print("sum = ", X_stat.sum())

sum =  11


In [275]:
# prod()
print("product = ", X_stat.prod())

product =  -720


#### Numpy - Mathematical and Statistical functions on NumPy Arrays - Sum(across axes)
<div class="aha-panel-body">
                    
<p>
    Please follow the following steps:</p>

<p>(1) Please import the required libraries</p>

<pre><code>import numpy as np
</code></pre>

<p>(2) Please create a NumPy array called <code>Z</code>, with a total of 18 elements, and shape as (2, 3, 3)</p>

<pre><code>&lt;&lt;your code comes here&gt;&gt; = np.arange(&lt;&lt;your code comes here&gt;&gt;).reshape(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p>(3) Print the above array <code>Z</code> using the <code>print()</code> function to see its values.</p>

<pre><code>print(Z)
</code></pre>

<p>(4) Please find the sum across <code>axis=2</code> for the array <code>Z</code>, and print the result using the <code>print()</code> function</p>

<pre><code>  print( Z.sum(&lt;&lt;your code comes here&gt;&gt;))
</code></pre>

</div>


In [287]:
Z = np.arange(18).reshape(2,3,3)
print(Z)

[[[ 0  1  2]
  [ 3  4  5]
  [ 6  7  8]]

 [[ 9 10 11]
  [12 13 14]
  [15 16 17]]]


In [288]:
  print( Z.sum(axis=2))

[[ 3 12 21]
 [30 39 48]]


#### Numpy - Mathematical and Statistical functions on NumPy Arrays - Transpose

<div class="aha-panel-body">
                    
<p>Please follow the below steps:</p>

<p>(1) Please import the required libraries</p>

<pre><code>  import numpy as np
</code></pre>

<p>(2) Please create a Numpy array called <code>N</code>  of size 6 and shape as (3,2)</p>

<pre><code>&lt;&lt;your code comes here&gt;&gt; = np.arange(&lt;&lt;your code comes here&gt;&gt;).reshape(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p>(3) Print array <code>N</code> using the <code>print()</code> function to see its values</p>

<pre><code>print(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p>(4) Print the transpose of array <code>N</code></p>

<pre><code>print("Transpose = ", &lt;&lt;your code comes here&gt;&gt;)
</code></pre>

                    
</div>

In [294]:
N = np.arange(6).reshape(3,2)
print(N)

[[0 1]
 [2 3]
 [4 5]]


In [297]:
#Transpose of a matrix (with rank >= 2) interchanges the rows and columns of the matrix.

# We can get the transpose of a matrix (or array) by using the attribute T.
print("Transpose = ", N.T)

Transpose =  [[0 2 4]
 [1 3 5]]


#### Numpy - Broadcasting in NumPy Arrays
<div class="aha-panel-body">
                    
<p>
    Please follow the below steps:</p>

<p>(1) Please import the required libraries</p>

<pre><code>import numpy as np
</code></pre>

<p>(2) Please create a numpy array <code>X_broad</code> of shape (3,3) filled with ones</p>

<pre><code>&lt;&lt;your code comes here&gt;&gt; = np.ones(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p>(3) Print array <code>X_broad</code> using <code>print()</code> function to see its values</p>

<pre><code>print(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p>(4) Please create a numpy array <code>Y_broad</code> of size 3</p>

<pre><code>Y_broad = np.arange(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p>(5) Print array <code>Y_broad</code> using <code>print()</code> function to see its values</p>

<pre><code>print(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p>(6) Add <code>X_broad</code> and <code>Y_broad</code> and store the result in a variable <code>Z_broad</code></p>

<pre><code>&lt;&lt;your code comes here&gt;&gt; = &lt;&lt;your code comes here&gt;&gt;  + &lt;&lt;your code comes here&gt;&gt;
</code></pre>

<p>(7) Print array <code>Z_broad</code> using <code>print()</code> function to see its values</p>

<pre><code>print(&lt;&lt;your code comes here&gt;&gt;)
</code></pre>

<p>Here you will observe that X_broad and Y_broad are added by the second rule of Broadcasting.</p>

</div>

In [305]:
X_broad = np.ones([3,3])
print(X_broad)

[[1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]]


In [306]:
Y_broad = np.arange(3)
print(Y_broad)

[0 1 2]


In [307]:
Z_broad = X_broad + Y_broad
print(Z_broad)

[[1. 2. 3.]
 [1. 2. 3.]
 [1. 2. 3.]]


<div class="aha-panel-body">
<h1 class="h2">
Numpy and Pandas - What is NumPy and Pandas ?
</h1>
					
<p>NumPy and Pandas are the Python libraries that are used to manipulate, process and analyze the data. They don't have constructs that can be used to visualize the data, for that we can use another library from Python called matplotlib.</p>

<h3>NumPy</h3>

<ul>
<li><p>NumPy stands for 'Numeric Python' or 'Numerical Python'.</p></li>
<li><p>It is designed for scientific computations.</p></li>
<li><p>It has efficiently implemented multi-dimensional arrays and it also provides fast mathematical functions.</p></li>
<li><p>It is mostly used for array-oriented computing.</p></li>
<li><p>NumPy's main object is the homogeneous multidimensional array called "ndarray".</p></li>
</ul>

<h3>Pandas</h3>

<ul>
<li><p>Pandas provide high-performance, easy-to-use data structures, and data analysis tools.</p></li>
<li><p>The DataFrame is the main and widely used data structure of the Pandas library.</p></li>
<li><p>DataFrame is a kind of in-memory 2-D table (similar to Excel sheet) with rows and columns. </p></li>
<li><p>Using DataFrames, we can create pivot tables (just like Excel sheet), compute one column value using values of other columns, etc.</p></li>
</ul>

<h3>When do we use NumPy and when we use Pandas?</h3>

<p>Suppose, your data is currently lying in a text file or a database table. You can load this dataset either using Pandas or Numpy. Pandas library is preferred as it is more efficient.</p>

<p>You can load the dataset using Pandas into a Pandas Dataframe. After loading the dataset, you can use Pandas library functions along with Matplotlib library functions to analyze, visualize and perform statistical analysis on the data in the dataset.</p>

<p>You can use NumPy's functions to create random indices and use these random indices with Pandas dataframe to create a shuffled dataset.</p>

<p>Many functions of the Scikit Learn (sklearn) library (like <code>Imputer</code>, <code>OneHotEncoder</code>, <code>predict()</code>) return a NumPy array, which we may need to process using NumPy. Also, we may need to create a Pandas dataframe from an existing NumPy array.</p>

<p><code>Imputer</code> - It is used to handle missing values in a dataframe
<br>
<code>OneHotEncoder</code> - It is used to perform one hot encoding on categorical values
<br>
<code>predict()</code> - It is used to predict values using a trained ML model</p>

<p>So, we see that Pandas and NumPy are used very much in conjunction with each other throughout the Machine Learning process.</p>

					
</div>